commit 35d851df23b093ee027f827fed2213ae5e88fc7a
Author: Jiri Kosina <jkosina@suse.cz>
Date:   Thu Aug 25 14:21:37 2011 +0200

    HID: magicmouse: ignore 'ivalid report id' while switching modes, v2
    
    This is basically a more generic respin of 23746a6 ("HID: magicmouse: ignore
    'ivalid report id' while switching modes") which got reverted later by
    c3a492.
    
    It turns out that on some configurations, this is actually still the case
    and we are not able to detect in runtime.
    
    The device reponds with 'invalid report id' when feature report switching it
    into multitouch mode is sent to it.
    
    This has been silently ignored before 0825411ade ("HID: bt: Wait for ACK
    on Sent Reports"), but since this commit, it propagates -EIO from the _raw
    callback .
    
    So let the driver ignore -EIO as response to 0xd7,0x01 report, as that's
    how the device reacts in normal mode.
    
    Sad, but following reality.
    
    This fixes https://bugzilla.kernel.org/show_bug.cgi?id=35022
    
    Reported-by: Chase Douglas <chase.douglas@canonical.com>
    Reported-by: Jaikumar Ganesh <jaikumarg@android.com>
    Tested-by: Chase Douglas <chase.douglas@canonical.com>
    Tested-by: Jaikumar Ganesh <jaikumarg@android.com>
    Signed-off-by: Jiri Kosina <jkosina@suse.cz>

Index: linux-2.6.37/drivers/hid/hid-magicmouse.c
===================================================================
--- linux-2.6.37.orig/drivers/hid/hid-magicmouse.c
+++ linux-2.6.37/drivers/hid/hid-magicmouse.c
@@ -532,10 +532,18 @@ static int magicmouse_probe(struct hid_d
 		goto err_stop_hw;
 	}
 	report->size = 6;
-
+ 
+	/*
+	 * Some devices repond with 'invalid report id' when feature
+	 * report switching it into multitouch mode is sent to it.
+	 *
+	 * This results in -EIO from the _raw low-level transport callback,
+	 * but there seems to be no other way of switching the mode.
+	 * Thus the super-ugly hacky success check below.
+	 */
 	ret = hdev->hid_output_raw_report(hdev, feature, sizeof(feature),
 			HID_FEATURE_REPORT);
-	if (ret != sizeof(feature)) {
+	if (ret != -EIO && ret != sizeof(feature)) {
 		dev_err(&hdev->dev, "unable to request touch data (%d)\n",
 				ret);
 		goto err_stop_hw;
